src/GitTools.jl

```diff
@@ -0,0 +1,125 @@
+module GitTools
+
+using ..Pkg3
+import LibGit2
+using Printf
+
+Base.@kwdef mutable struct MiniProgressBar
+    max::Float64 = 1
+    header::String = ""
+    color::Symbol = :white
+    width::Int = 40
+    current::Float64 = 0.0
+    prev::Float64 = 0.0
+    has_shown::Bool = false
+end
+
+function showprogress(io::IO, p::MiniProgressBar)
+    perc = p.current / p.max * 100
+    prev_perc = p.prev / p.max * 100
+    # Bail early if we are not updating the progress bar,
+    # Saves printing to the terminal
+    if p.has_shown && !(perc - prev_perc > 0.1)
+        return
+    end
+    p.prev = p.current
+    p.has_shown = true
+    n_filled = ceil(Int, p.width * perc / 100)
+    n_left = p.width - n_filled
+    print(io, "    ")
+    printstyled(io, p.header, color=p.color, bold=true)
+    print(io, " [")
+    print(io, "="^n_filled, ">")
+    print(io, " "^n_left, "]  ", )
+    @printf io "%2.1f %%" perc
+    print(io, "\r")
+end
+
+Base.@kwdef struct GitTransferProgress
+    total_objects::Cuint
+    indexed_objects::Cuint
+    received_objects::Cuint
+    local_objects::Cuint
+    total_deltas::Cuint
+    indexed_deltas::Cuint
+received_bytes::Csize_t
+end
+
+function transfer_progress(progress::Ptr{GitTransferProgress}, p::Any)
+    progress = unsafe_load(progress)
+    bar = unsafe_pointer_to_objref(p.transfer_progress)
```

Comment:
I would have assumed that the payload sent to this function was supposed to be the `p.transfer_progress` instead of the whole struct.

Reply:
I forgot to update this code with my Pkg3.jl PR. This will need to be changed to:
```julia
transfer_progress_ptr = payload_unpack(p, :transfer_progress)
@assert transfer_progress_ptr != C_NULL
bar = unsafe_pointer_to_objref(transfer_progress_ptr)
```
At the moment if you use the tuple syntax for `RemoteCallbacks` it breaks apart the tuple and puts each of the payloads inside of a `RemotePayloads` container. I decided that having each callback unpack this object was the better solution instead of unwrapping it before calling the user's callback.

Reply:
>  I decided that having each callback unpack this object was the better solution instead of unwrapping it before calling the user's callback.

Alright.

Reply:
@omus 

> payload_unpack(p, :transfer_progress)

doesn't work because `payload_unpack` is an instance of `RemotePayloads` and not a pointer to it (like `payload_unpack` expects).